### DISCUSSION TITLE
2024-08-04 회의록

### DISCUSSION THREAD

**Discussion by Captain-Kim** [nbc-final-a10/TripBuddy#193] [회의록]
### ⭐️ 체크리스트

- [x] : 개인 별 기능 구현이 다 되었는가? 개인 별로 직접 최신 dev로 화면공유 시연 (To.everyone)
- [x] : 공통 기술 문서 작성 (To.병준)
- [x] : UT 기획 내용 브리핑 (To.유상)
- [x] : 개인별 스크립트 작성(댓글) (To.everyone) (회의 대화 끝나는 대로 바로 요거부터 부탁드립니다)
- [x] : 더미 데이터 제대로 만들기 (프사나 정보 없는 유저 X, 진짜 유저가 쓴 것 같은 여정 모집 글) (To.everyone) (본인 온보딩한번씩 /onboarding에서 및 글 한개씩 작성 및 다른분이 만든거에 참여참여 누르기 - 내일까지)
- [x] : 최종 merge time (지금 안 되면 멕시멈 19시) (To.everyone) (오은/김민지/곽현정 - PR 주말 핫픽스로 처리)
- [x] : PPT (To.현정) ( 기술적 의사결정 / 트러블 슈팅 / 추후 개발 및 기술적 도전 계획 - 붉은색 처리된 곳들)
- [x] : 미들웨어 처리 
- [x] : 최종 dev to main 테스트 (To.everyone) (오늘 저녁/밤에 어프로브없이 바로 진행)
- [x] : 발표 대본 (To.은) (월요일 10시 경 모의발표)

### ⭐️ 미들웨어 체크리스트
- 퍼블릭/컨디셔널은 안튕기고 어센티케이트드는 튕기고
- 라이트페이지/ 비인가는 접근 금지
- 채팅/ 
- 미인증 회원은 메인이 아니라 튜토리얼로 리다이렉트

### ⭐️ 개인 별 핵심 기능 및 트러블 슈팅 작성

- 서식

```
## 작성자 : 김병준

### 맡은 페이지(기능)

- 글쓰기 페이지(CUD)

### 핵심 기술

- Funnel Pattern

### 트러블 슈팅

- Funnel Pattern을 여러 페이지에서 재사용 가능하도록 설계하는 데 어려움이 있었습니다.

### 잘 된 점, 미흡한 점

#### 잘 된 점

- 함수 및 컴포넌트의 재사용성이 높은 팀프로젝트입니다.

#### 미흡한 점

- 최적화가 부족합니다.

### 개선 과제(도전 과제)

- 아토믹 패턴으로 재설계가 필요해보입니다.

### 시연 유의 사항

#### 강조할 사항

- 글쓰기 페이지 시연 시 Funnel Pattern에 대해 강조해주세오.
- 마이페이지에서는 온도 지수에 대해 강조해주세요.

#### 주의해야 할 사항

- [마이페이지] : 반응형 작업이 안 되어 있으니 PC 화면으로 늘리지 말아주세요.

```

- 아래 댓글에 작성해주세요.

## 🍀 MVP 발표 체점표

- [x] : 1. 프로젝트 서비스 기획에 따라 에러나 버그없이 웹사이트가 잘 동작하는가? **=> 일부 있으나 기획 의도와 다른 것은 아님**
- [x] : 2.  MVP로 기획한 기능을 모두 개발했는가? **=> 개발 자체는 완료**
- [x] : 3. 팀원들끼리 정한 디렉토리 구성 규칙이 있는지? 규칙이 있다면 해당 규칙에 맞게 잘 지켰는가? **=> 아키텍처 설계 잘 **
- [x] : 4. 개발에 활용한 프레임워크/라이브러리 선택에 대한 기준이 명확한가?  (ex. CRA, React-Vite, Next.js 중 어떤 프레임워크를 사용했고 그 이유는?) **=> (Next.js) 서버 컴포넌트를 혼용하여 SEO 최적화하기 위해**
- [x] : 5. 작성한 코드의 가독성이 좋고 의도가 명확한가?  (변수명의 semantic 여부, 1개 함수에는 1개의 기능 등) **=> 1개의 함수에서 다양한 로직을 구현하는 게 더 나은 경우도 많은 듯**
- [x] : 6. 버전관리시스템(Git)을 잘 활용하여 협업 및 코드 관리 했는가?  (commit, PR, 리뷰, 브랜치 관리정책 등) **=> Projects, Discussion 등을 잘 활용하였고 PR 컨벤션도 모두 준수하였으며 브랜치 전략 또한 명확함**
- [x] : 7.  컴포넌트와 커스텀훅 등을 사용하여 코드의 재사용성을 고려했는가? **=> 이것 때문에 400 커밋을 함**
- [ ] : 8. React의 Virtual DOM의 원리를 고려하지 않은 코드가 있는가?  (Vanilla JS 레퍼런스 코드를 그대로 적용하진 않았는가?) **=> state 사용 등이 적절하고 생애주기를 고려한 useEffect 훅 사용도 적절함**
- [x] : 9. 트러블 슈팅의 접근과 해결 과정을 명확히 설명할 수 있는가? **=> 개별 작성**
- [x] : 10. 반응형 UI 까지 고려했는가? (+@) **=> 커리큘럼 상 모바일, PC 시안 중 하나만 나왔는데 어떻게? 그럼에도 우리는 처음부터 했음**
- [x] : 11. React.memo, useMemo, useCallback 등을 활용하여 불필요한 컴포넌트 렌더링을 최소화 하였는가? **=> (useMemo) slate editor에서 `createEditor`로 인스턴스를 생성해야 하는데 비제어컴포넌트가 리렌더링 될 때마다 인스턴스가 재생성 되지 않도록 에디터를 메모이제이션 함. (useCallback) 네이버 로그인, 커스텀 모달 오픈, 퍼널 버튼 스텝 증가 등에서 함수의 재생성을 막기 위해 무수히 사용됨**
- [x] : 12. (TS를 사용한 경우) any type 없이 엄격한 타입 제어를 통해 발생 가능한 오류를 미연에 방지하였는가? **=> currentBuddy나 PWA 등 아직 손 안 댄 부분 빼고는 any로 캐스팅 한 곳 아무곳도 없음**
- [x] : 13. (TS를 사용한 경우) enum type 또는 공통타입 코드 등을 사용하여 하드코딩을 최소화 하였는가? **=> enum을 쓸 상황이 없었음**
- [x] : 14. (SUPABASE 사용한 경우) 테이블 구조를 왜 그렇게 설계했는지를 명확히 설명할 수 있는가? (Schema Visualizer (ERD) 기반 설명) **=> supabase 전문가가 설계함, 비주얼라이저를 별도 앱에서 작성하고 SQL로 테이블을 셋업할 정도로 수준 높게 작성함. 또한 팔로워 수 등을 계산하기 위한 하드코딩을 피하기 위해 SQL문도 적극 활용함**
- [x] : 15. (React query 사용한 경우) 라이프사이클을 이해하여 query key를 통한 캐싱 처리를 적절하게 하였는가? => **쿼리 키만 따로 모아놓을 정도로 관리 잘 함**
- [x] : 16. 기술멘토님의 질문에 CS지식이나 합리적인 근거에 따라 답변 하고 있는가? **=> 그럴 예정**

// 이미지 테이블
<table>
  <tr>
    <th>이미지 1번 제목</th>
    <th>이미지 2번 제목</th>
  </tr>
  <tr>
    <td><img src="이미지 주소" width="300" height="300"/></td>
    <td><img src="이미지 주소" width="300" height="300"/></td>
  </tr>
</table>

**Comment by Captain-Kim** [comment]
### 맡은 페이지(기능)

- 글쓰기 페이지(CUD)
- 마이 페이지(팔로우, 온도지수)
- 랭킹 페이지

### 핵심 기술

- Funnel Pattern
- Rich-Text-Editor(Toolbar customizing, Image base-64 decoding)

### 트러블 슈팅

- Funnel Pattern을 여러 페이지에서 재사용 가능하도록 설계하는 데 어려움이 있었습니다.
  - 페이지마다 버튼의 라벨이 다를 수 있다는 점, 페이지 마다 버튼 클릭 시 필요한 액션이 다를 수 있다는 점, 페이지 마다 필요한 버튼 스타일이 다를 수 있다는 점 등을 각 페이지가 제작되는 과정에서 계속 발견되어 유지보수하였습니다.
  - 계속 수정하는 과정에서 이 패턴을 사용하는 팀원의 코드에서 충돌이 나지 않도록 꼭 필요한 props와 그렇지 않은 props를 충분히 고민하여 유지보수해야 하는 과정이 필요했습니다.
  - 또한 나만 이해할 수 있는 코드가 아니라 다른 팀원들도 직관적으로 이해할 수 있는 로직으로 사용해야 팀원들이 쉽게 사용할 수 있기에, 품질 높은 코드를 작성하기 위해 노력했고 github에 기술 문서를 작성하는 등 재사용성을 높이기 위한 다양한 방법을 시도했습니다.

### 잘 된 점, 미흡한 점

#### 잘 된 점

- 함수 및 컴포넌트의 재사용성이 높은 팀프로젝트입니다.
  - 예상보다 기능 구현이 늦어졌는데, 처음 도착한 와이어프레임을 보았을 때 대부분은 거의 비슷한데 완전히 같거나 재사용이 가능해보이는 컴포넌트가 아니었습니다.
  - 팀원들이 와이어프레임을 며칠 간 분석하면서 재사용이 가능한 컴포넌트를 찾아 담당자를 명확히 하였고, 각기 다른 페이지에서 컴포넌트나 함수를 호출했을 때 발생할 사이드이펙트까지 고려하며 mode를 props로 내리는 등 불필요한 코드 작성이 zero에 수렴할 정도로 재사용을 하였습니다.

#### 미흡한 점

- 최적화 기법을 잘 몰라 코드를 재사용하거나 불필요한 자바스크립트를 줄이는 등의 리팩토링을 제외하고는 별도의 최적화가 이루어지지 않았습니다.
- Next.js로 제작한 프로젝트에서 매번 느린 느낌을 받았었고, 이번 프로젝트는 특히 프로젝트의 규모가 커 시작하기 전부터 걱정한 것에 비해서는 속도가 괜찮지만 아직도 서비스로 내놓기에는 한참 부족하다고 느낍니다. 최적화가 필요합니다.

### 개선 과제(도전 과제)

- 아토믹 패턴으로 아키텍처를 디자인 하였지만, 정말 제대로 사용하고 있는 지에 대해서는 심도있는 재설계가 필요할 것 같습니다. atoms로 분리한 것을 다른 컴포넌트에서도 재사용한다면 페이지마다 일관성 있는 UI를 만들 수도 있을 것 같은데, 시간 부족으로 처음부터 atoms의 재사용까진 설계하지 못하고, 제작한 팀원만이 atoms를 이해하고 있습니다. 허나 이것은 물리적으로 프로젝트 기간이 부족해 불가능 해 보이지만 실제 서비스라면 이런 과정도 필요해보입니다.

### 시연 유의 사항

#### 강조할 사항

- 글쓰기 페이지 시연 시 Funnel Pattern에 대해 강조해주세오.
  - 사용자에게 받아야 하는 정보가 많아 한 페이지로 구성하지 않고 약 7개의 컴포넌트로 구성하여 시퀀스가 이동함에 따라 사용자는 마치 하나의 페이지를 이용하는 것 같은 매끄러운 경험을 할 수 있습니다.
  - 초기 로딩 시간이 소폭 상승한 점을 제외하면, 이후부터는 페이지 이동 속도가 zero이기 때문에 사용자는 페이지 전환이 매우 빠르다고 느낄 수 있습니다.
  - Funnel Pattern은 글쓰기 페이지 뿐만 아니라 온보딩, 튜토리얼에서도 사용되는 디자인 패턴으로 재사용이 가능하도록 로직을 설계하였습니다. 이에 사용하는 페이지에서 자유롭게 step의 개수를 조절할 수 있고, 버튼 라벨 또한 모두 다르게 설정이 가능합니다. 또한 특정 step에서 버튼 클릭 시 유효성 검사가 필요하거나 api 통신을 해야 할 때 필요한 함수를 사용할 수도 있습니다.
- 마이페이지에서는 온도 지수에 대해 강조해주세요.
  - 다만 온도 지수는 구현이 되어 있지만 사용자가 글을 작성하거나 팔로워가 증가할 때 등, 알고리즘에 의해 증가, 감소하는 로직을 구현할 예정입니다.
  - 이를 통해 사용자들의 서비스 이용을 독려할 수 있습니다.

#### 주의해야 할 사항

- **[마이페이지]** : 반응형 작업이 안 되어 있으니 PC 화면으로 늘리지 말아주세요.
- **[팔로우]** : 구현이 안 되어 있으니 누르지 말아주세요.
- **[글쓰기 페이지]** : 모든 항목에 유효성 검사가 들어가있지 않으니 3개 선택해야 하는 chips 등에서 놓치지 말고 모두 선택해주세요. null이 들어가면 글쓰기에 실패할 수 있습니다.
- **[글쓰기 페이지]** : 고해상도 사진을 넣으면 글쓰기에 실패할 수 있습니다. null 또는 통신 장애로 글쓰기에 실패해도 성공 step이 떠서 이상해집니다.

**Comment by eunohhh** [comment]
### 맡은 페이지(기능)

- 인증 / 인가
- 스토리 페이지
- 여정 페이지

### 핵심 기술

- supabase auth
- naver social login

### 트러블 슈팅

- 팀원들이 같이 사용할 수 있도록 재사용 가능한 컴포넌트를 만드는 것이 여러 조건을 고려해야 하다보니 생각보다 까다로웠습니다.
- 네이버 소셜 로그인 구현시 supabase auth 에 기본 지원하지 않아 약간 수동적인 방법으로 작성하다 보니 여러움이 있었습니다.

### 잘 된 점, 미흡한 점

#### 잘 된 점

- 꾸준한 회의를 통해 요구사항과 담당임무를 명확히 하고 이를 바탕으로 작업하는 방식이 높은 효율성을 보여주었습니다.

#### 미흡한 점

- 최적화 문제
- 속도 문제
- 라이트하우스
- 인피니트 쿼리

### 개선 과제(도전 과제)

- 전체적인 페이지의 최적화와 라이트하우스 점수 개설

### 시연 유의 사항

#### 강조할 사항

- 인증 / 인가 중에서는 네이버 로그인만 보여주세요
- 스토리 기능이 얼마나 완성될지 모르겠지만, 보여주면 좋을 것 같습니다.
 
#### 주의해야 할 사항

- **[스토리페이지]** : 미완성입니다.

**Comment by Kyara0324** [comment]
## User Test 질문 리스트

### 1. 사용자 경험
- 앱을 처음 사용했을 때, 사용하기 쉽다고 느꼈나요?
- 첫 화면에서 어떤 기능이 가장 먼저 눈에 들어왔나요?
- 기능들을 찾고 사용하는 데 어려움이 있었나요?
- 전체적인 UI 디자인이 만족스러웠나요?

### 2. 기능성
- 가장 유용하다고 생각한 기능은 무엇인가요?
- 가장 불편하거나 개선이 필요한 기능은 무엇인가요?
- 특정 기능이 제대로 작동하지 않거나 오류가 발생한 적이 있나요?
- 사용 중 느린 반응 속도를 경험한 적이 있나요?

### 3. 신뢰성과 안전성
- 데이터가 안전하게 보호된다고 느껴졌나요?
- 로그인 및 회원가입 과정에서 문제를 겪었나요?
- 오류나 버그를 발견했을 때 이를 쉽게 보고할 수 있었나요?

### 4. 전반적인 만족도
- 전체적인 만족도를 1점에서 10점으로 평가해 주세요.
- 다시 이 앱을 사용할 의향이 있나요?
- 친구나 가족에게 추천할 의향이 있나요?
- 개선되었으면 하는 부분이 있나요?

#### +팀원들에게 맡은 부분 질문한 내용


===================작성해주실 내용===================

### 1. 프론트엔드 팀원 (전체)
- UI 디자인에서 사용자가 가장 쉽게 이해할 수 있는 부분은 어디라고 생각하나요?
- 반응형 디자인을 구현하면서 가장 신경 쓴 부분은 무엇인가요?
- 사용자 피드백을 바탕으로 어떤 부분을 가장 먼저 개선하고 싶나요?

### 2. 백엔드 팀원 (전체)
- 데이터베이스 구조에서 가장 중점을 둔 부분은 무엇인가요?
- API를 설계할 때 가장 중요하게 고려한 점은 무엇인가요?
- 현재 시스템의 안정성을 위해 어떤 방안을 마련했나요?

### 3. QA 팀원 (유상)
- 테스트 과정에서 가장 많이 발견된 버그는 무엇이었는지
- 사용자 시나리오 테스트에서 주로 어떤 부분에 집중했는지
- 사용자 피드백을 통해 발견된 문제점 중 가장 중요한 것은 무엇이었는지

**Comment by slsl2** [comment]
## 작성자 : 김민지

### 맡은 페이지(기능)

- home 페이지 스켈레톤, 배너
- 채팅 목록 페이지
- 채팅 페이지 실시간 채팅

### 핵심 기술

- realtime 실시간 채팅

### 트러블 슈팅

문제: supabase 테이블 여러가지 동시에 엮어서 데이터 연결.
그동안 한 페이지에서 테이블 한 두 개만 사용했는데, 각각 엮여 있는 네 개의 테이블에 접근해 페이지를 구성하는 것에 어려움을 겪었습니다.
필요한 테이블과 column을 정리하고, 코딩 전에 로직을 미리 정리한 다음 순차적으로 접근하는 방법으로 진행하였습니다.

### 잘 된 점, 미흡한 점

#### 잘 된 점

- 채팅 기능 구현

#### 미흡한 점

- 인가 처리가 미흡하다고 느끼고 있습니다.

### 개선 과제(도전 과제)

- MVP 이후 도전 과제: 실시간 알림 기능 구현

### 시연 유의 사항

#### 강조할 사항

- 채팅 탭 두개 켜서 시연
- postgres_changes 사용, useState 써서 message 테이블에 insert 이벤트가 발생할 때마다 seMessages 업데이트

#### 주의해야 할 사항

- **[채팅 페이지]** 현재 직접 URL 창에 /chat/[id] 입력하면 아무나 들어갈 수 있어서 (인가 처리 안 됨) 채팅 페이지는 꼭 채팅 목록에서 클릭해 들어가 주시면 좋을 것 같습니다.
- **[채팅 페이지]** 지금 채팅 입장 시점 전의 메시지도 전부 보이는데, 이 부분 개선할 예정이라고 말해도 될 것 같습니다.

**Comment by Kyara0324** [comment]
## 작성자 : 최유상

### 맡은 페이지(기능)

- 튜토리얼 페이지

### 핵심 기술

- Funnel Pattern / indicator 

### 트러블 슈팅

- 각 페이지마다 이미지 크기에 따라 ui 위치가 달라지는점과 글자깨짐현상이 발생하는 것에 어려움이 있었습니다.

### 잘 된 점, 미흡한 점

#### 잘 된 점

- 튜토리얼 페이지를 와이어프레임과 동일하게 만듦
- 팀원분들이 고생하며 만들어두신 컴포넌트들을 재사용하여 코드의 재사용성을 높임(?)

#### 미흡한 점

- 아직 tailwind에 대한 이해가 부족하여 원하는대로 짧은시간내에 딱딱 만들어내지 못하는점

### 개선 과제(도전 과제)

- 웹 버전에서 나타낼 튜토리얼 페이지를 렌더링하는 문제 ( 새로 만들지 여기서 조건부 렌더링으로 전체적인 ui를 키워 렌더링할지 )

### 시연 유의 사항

-너무 작은 핸드폰모델에서 시연시 스크롤이 생기는 문제가 발생해 사용자경험이 떨어진다는 느낌을 받을 수 있으므로 iPhone XR이상의 모델에서 시연해주세요.

#### 강조할 사항

- 튜토리얼 페이지 시연 시 Funnel Pattern에 대해 강조해주세요.
- 건너뛰기 버튼이 마지막페이지에서는 없어져 중복되는 기능이 없다는점을 강조해주세요.

#### 주의해야 할 사항

-너무 작은 핸드폰모델에서 시연시 스크롤이 생기는 문제가 발생해 사용자경험이 떨어진다는 느낌을 받을 수 있으므로 iPhone XR이상의 모델에서 시연해주세요.

-이미지를 자주 변경해 간혹 이전버전의 이미지가 출력되는 현상이 발생할 수 있으므로 최신버전의 dev branch를 pull받아 시연해주세요.

**Comment by hyunjeongkwak** [comment]
## 작성자 : 곽현정

### 맡은 페이지(기능)

- 검색 페이지

### 핵심 기술

- 검색 filter 기능

### 트러블 슈팅

- zustand로 코드를 작성하다가 props를 내려받는 것으로 리팩토링을 하는 과정이 까다로웠습니다.
- 다중 조건이 걸린 필터 기능을 구현하는 부분이 까다롭다고 느껴졌습니다.

### 잘 된 점, 미흡한 점

#### 잘 된 점

- 컴포넌트의 재사용성이 높아 그 부분이 잘된 부분이라고 생각합니다.

#### 미흡한 점

- 속도가 느린 부분이 아쉽다고 느껴졌습니다.

### 개선 과제(도전 과제)

- 연관 검색어, 자동 완성 기능

### 시연 유의 사항

#### 강조할 사항

- 원하는 옵션을 선택하면 해당하는 모든 데이터가 가로 스크롤로 렌더링되고,
- 검색 결과에 포함되지 않은 여정은 'trip_start_date'의 데이터를 기준으로 여정 임박순 정렬하여 추천 여정에 뜹니다.

#### 주의해야 할 사항

- **[검색 페이지]**: 여정 테마, 버디즈 성향 선택 옵션은 기능 구현이 되어 있지 않습니다.
- xl ui가 아직 제대로 구현되지 않아, 검색 옵션 부분은 pc버전으로 늘리지 않고 시연하는 것을 권장합니다.
